### PR TITLE
Slasher Significant Optimizations

### PR DESCRIPTION
**Pull request by rauljordan**
# Background

This PR contains significant optimizations to Prysm's slasher implementation, recently released in [v2.0.2](https://github.com/prysmaticlabs/prysm/releases/tag/v2.0.2) as a result of a deep comparison between the Lighthouse team's implementation of slasher both in benchmarks using real data and also in the Prater testnet.

It is clear that Prysm's slasher is performing poorly from users running it, sometimes leading to incredibly high memory usage. Additionally, Prysm's slasher sometimes takes ages to run on a group of valid attestations, making us suspect the runtime is getting clogged up due to the slow processing, and making it hard to recover and produce reliable results.

## Current Testnet Performance

Prysm currently performs slashable attestations checks in around 7-14 seconds, sometimes taking **7 minutes** to process a group of valid attestations. The variance is astounding. Lighthouse consistently processes 1000 attestations per slot in **1.5 seconds** on average. Prysm's memory footprint is huge, with some users reporting maxing out RAM.

## Local Benchmark with Real Fixtures

The following tests were performed on a M1Pro Chip ARM Darwin MacbookPro 2021 using Go 1.17 and the stable release of Rust as of Oct 28, 2021.

### In Prysm

To really understand why we have this huge discrepancy and poor performance in Prysm, I extracted all attestations from two blocks in [mainnet](https://beaconcha.in) testnet, along with all beacon committees for its epoch to convert the attestations to indexed form. The attestations were then written as SSZ encoded bytes into a local output directory. For ease of reproduction, a zip file of the indexed attestations folder is attached here below from slots 2385621 and 2385622:

[attestations.zip](https://github.com/prysmaticlabs/prysm/files/7436955/attestations.zip)

and

[attestations2.zip](https://github.com/prysmaticlabs/prysm/files/7437137/attestations2.zip)

The data was extracted using a bespoke Go tool called `fetch-attestations` and can be seen in the diff below:

https://github.com/prysmaticlabs/prysm/compare/develop...slasher-benching

Additionally, I began writing a local test that reads these attestations and passes them through our slasher while logging out processing times. What was found was not ideal. 

**Current Prysm Slasher**
```
Got a snapshot of valid attestations 128
Took 14 minutes, 980 batches by validator chunk index
```
**14 minutes** to process a simple block of attestations. Something must be wrong...

### Using Lighthouse

In a fork of the lighthouse repo, I wrote rust tests to perform the same local benchmark with the same fixtures through their slasher implementation:

https://github.com/sigp/lighthouse/compare/stable...rauljordan:stable

On rust stable, `cd slasher && cargo test -- --nocapture`

**Current Lighthouse Slasher**

```
Reading attestations from slot N+1
Processing queue
Got a snapshot of valid attestations 128
Storing valid attestations in slasher DB
Grouping attestations by validator chunk index
Processing 980 batches
test process_attestations_from_file has been running for over 60 seconds
Took 4.03 minutes, avg 247ms per batch

Reading attestations from slot N+1
Processing queue
Got a snapshot of valid attestations 128
Storing valid attestations in slasher DB
Grouping attestations by validator chunk index
Processing 980 batches
Took 10.7 minutes, avg 655ms per batch
```

Clearly we are much slower than lighthouse, and it seem to perform far better at different steps of the process.

## Optimizations

Upon further investigation, here are the problems found and how they were solved:

**_We group attestations into batches by validator chunk indices, but we perform double vote detection on each batch_**

Double vote detection is a fairly expensive part of the slasher routine, but we were overdoing it for every single attestation batch. Instead, we could do it only **once** for the entire list of valid attestations whilst processing.

**_We were updating the entire validator registry's latest written epoch in a single BoltDB transaction_**

We had a bolt transaction that was storing _250,000 entries_ all at once, during each iteration of slashing detection over attestations. This is extremely excessive and too much for boltDB to handle. We believe this is a root cause of the excessive memory use, as bolt hates extremely large and long-running transactions. Instead, we only update the validators we care about and avoid doing so in a giga-sized DB transaction.

**_We were neither pre-fetching nor updating validator min/max span chunks for the change in current epoch_**
Loading chunks from disk is an expensive procedure. However, all chunks need to be updated for validator indices in attestations if there is a change in epoch. We can use the start of slashing detection to load and update the relevant chunks for the change in epoch, and this allows us to **cache** loaded chunks for later use and fewer DB reads

**_We were updating validator chunk data for validators that are irrelevant to the attestation in question_**

Screen Shot 2021-10-28 at 3 52 06 PM

We had some code that had a map of latest epochs written by validator index. For nil entries, we would set the latest epoch written to `0`. Our code that would perform the update would consider this as an epoch to update from, leading to incredibly inefficient and unnecessary updates to validator chunks _from genesis_ when they could have been skipped altogether. Instead, we add an `if ok := someMap[validatorIndex]` check to the map before doing further processing, avoiding an extremely slow operation.

## Final Results

Prysm now **matches** Lighthouse's performance on mainnet:
```
level=info msg="Done checking slashable attestations" avgBatchProcessingTime=1.669984ms elapsed=1.670214893s numAttestations=900 numBatchesByValidatorChunkIndex=1000 prefix=slasher
```

### CODE REVIEW

File: beacon-chain/slasher/detect_attestations.go [hunk 1]
```diff
@@ -3,40 +3,67 @@ package slasher
...
+	start := time.Now()
+	doubleVoteSlashings, err := s.checkDoubleVotes(ctx, atts)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not check slashable double votes")
+	}
+	log.WithField("elapsed", time.Since(start)).Info("Done checking double votes")
+	slashings = append(slashings, doubleVoteSlashings...)
+
 	groupedAtts := s.groupByValidatorChunkIndex(atts)
+	log.WithField("numBatches", len(groupedAtts)).Info("Batching attestations by validator chunk index")
```
Comment (prestonvanloon) [inline beacon-chain/slasher/detect_attestations.go:33]: Can this go to debug?

File: beacon-chain/slasher/detect_attestations.go [hunk 2]
```diff
@@ -3,40 +3,67 @@ package slasher
...
+		avgProcessingTimePerBatch = avgProcessingTimePerBatch / time.Duration(len(batchTimes))
+	}
+	log.WithFields(logrus.Fields{
+		"numAttestations":                 len(atts),
+		"numBatchesByValidatorChunkIndex": len(groupedAtts),
+		"elapsed":                         time.Since(start),
+		"avgBatchProcessingTime":          avgProcessingTimePerBatch,
+	}).Info("Done checking slashable attestations")
+	if len(slashings) > 0 {
+		log.WithField("numSlashings", len(slashings)).Info("Slashable attestation offenses found")
```
Comment (prestonvanloon) [inline beacon-chain/slasher/detect_attestations.go:66]: What about bumping this to WARN level? It's an abnormal log that should attract users' attention.

File: beacon-chain/slasher/service.go [hunk 3]
```diff
@@ -144,6 +144,18 @@ func (s *Service) Stop() error {
 	if s.pruningSlotTicker != nil {
 		s.pruningSlotTicker.Done()
 	}
+	// Flush the latest epoch written map to disk.
+	start := time.Now()
+	log.Info("Flushing last epoch written for each validator to disk")
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:149]: Debug?

File: beacon-chain/slasher/service.go [hunk 4]
```diff
@@ -144,6 +144,18 @@ func (s *Service) Stop() error {
...
+	if err := s.serviceCfg.Database.SaveLastEpochsWrittenForValidators(
+		s.ctx, s.latestEpochWrittenForValidator,
+	); err != nil {
+		log.Error(err)
+	}
+	log.WithField("elapsed", time.Since(start)).Info(
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:155]: Debug?

File: beacon-chain/slasher/service.go (hunk 3 above)
Comment (rauljordan) [reply beacon-chain/slasher/service.go]: I want to keep this one because I want users to wait and not forcefully shut down their machine

**Review by prestonvanloon — DISMISSED** [review]

**Review by prestonvanloon — APPROVED** [review]

File: beacon-chain/slasher/service.go [hunk 5]
```diff
@@ -152,8 +164,44 @@ func (s *Service) Status() error {
 	return nil
 }
 
+func (s *Service) waitForChainInitialization() {
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := s.serviceCfg.StateNotifier.StateFeed().Subscribe(stateChannel)
+	defer stateSub.Unsubscribe()
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:170]: Do you need to do this?
```suggestion
	defer stateSub.Unsubscribe()
	defer close(stateChannel)
```

File: testing/slasher/simulator/simulator.go [hunk 6]
```diff
@@ -260,6 +264,8 @@ func (s *Simulator) verifySlashingsWereDetected(ctx context.Context) {
 				"prevTargetEpoch": slashing.Attestation_2.Data.Target.Epoch,
 				"sourceEpoch":     slashing.Attestation_1.Data.Source.Epoch,
 				"prevSourceEpoch": slashing.Attestation_2.Data.Source.Epoch,
+				"prevData":        fmt.Sprintf("%#x", slashing.Attestation_1.Data.BeaconBlockRoot),
+				"newData":         fmt.Sprintf("%#x", slashing.Attestation_2.Data.BeaconBlockRoot),
```
Comment (prestonvanloon) [inline testing/slasher/simulator/simulator.go:268]: Can you choose a better name for these fields? Data is ambigous.

File: beacon-chain/slasher/service.go [hunk 7]
```diff
@@ -49,77 +49,78 @@ type SlashingChecker interface {
...
-		log.WithField("genesisTime", s.genesisTime).Info("Starting slasher, chain already initialized")
-	} else {
-		// This should not happen.
-		log.Error("Could start slasher, could not receive chain start event")
+	log.Info("Completed chain sync, starting slashing detection")
+
+	// Get the latest eopch written for each validator from disk on startup.
+	headState, err := s.serviceCfg.HeadStateFetcher.HeadState(s.ctx)
+	if err != nil {
+		log.Error(err)
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:98]:
```suggestion
		log.WithError(err).Error("Failed to fetch head state")
```

File: beacon-chain/slasher/rpc_test.go [hunk 8]
```diff
@@ -125,7 +126,7 @@ func TestIsSlashableAttestation(t *testing.T) {
 		{
 			name:         "should detect multiple surround if multiple same indices",
 			attToCheck:   createAttestationWrapper(t, 1, 4, []uint64{0, 1}, []byte{2}),
-			amtSlashable: 2,
+			amtSlashable: 4,
```
Comment (prestonvanloon) [inline beacon-chain/slasher/rpc_test.go:129]: How can this be 4 if only two indices attested?

File: beacon-chain/slasher/receive.go [hunk 9]
```diff
@@ -218,9 +215,11 @@ func (s *Service) pruneSlasherDataWithinSlidingWindow(ctx context.Context, curre
 	if err != nil {
 		return errors.Wrap(err, "Could not prune proposals")
 	}
-	log.WithFields(logrus.Fields{
-		"prunedAttestationRecords": numPrunedAtts,
-		"prunedProposalRecords":    numPrunedProposals,
-	}).Info("Successfully pruned slasher data")
+	if numPrunedAtts > 0 || numPrunedProposals > 0 {
```
Comment (prestonvanloon) [inline beacon-chain/slasher/receive.go:218]: imo: always log this, whether something was pruned or not. It helps to see that pruning is working and completing task successfully.

Optional: Add the prune time duration as a field, that might look nice for debugging or performance monitoring. Other logs reference "elapsed" as field name.

File: beacon-chain/slasher/rpc_test.go (hunk 8 above)
Comment (rauljordan) [reply beacon-chain/slasher/rpc_test.go:129]: 4 slashable attestations were found. Two indices, each one has a surrounding and a surrounded attestation

Comment (rauljordan) [reply beacon-chain/slasher/rpc_test.go:129]: It's kind of weird, but yes slasher is technically finding 4 total attestations here

File: testing/slasher/simulator/simulator.go [hunk 10]
```diff
@@ -256,10 +260,12 @@ func (s *Simulator) verifySlashingsWereDetected(ctx context.Context) {
...
-				"targetEpoch":     slashing.Attestation_1.Data.Target.Epoch,
-				"prevTargetEpoch": slashing.Attestation_2.Data.Target.Epoch,
-				"sourceEpoch":     slashing.Attestation_1.Data.Source.Epoch,
-				"prevSourceEpoch": slashing.Attestation_2.Data.Source.Epoch,
+				"targetEpoch":         slashing.Attestation_1.Data.Target.Epoch,
+				"prevTargetEpoch":     slashing.Attestation_2.Data.Target.Epoch,
+				"sourceEpoch":         slashing.Attestation_1.Data.Source.Epoch,
+				"prevSourceEpoch":     slashing.Attestation_2.Data.Source.Epoch,
+				"prevBeaconBlockRoot": fmt.Sprintf("%#x", slashing.Attestation_1.Data.BeaconBlockRoot),
+				"newData":             fmt.Sprintf("%#x", slashing.Attestation_2.Data.BeaconBlockRoot),
```
Comment (prestonvanloon) [inline testing/slasher/simulator/simulator.go:268]:
```suggestion
				"newBeaconBlockRoot":  fmt.Sprintf("%#x", slashing.Attestation_2.Data.BeaconBlockRoot),
```

**Review by prestonvanloon — APPROVED** [review]

File: beacon-chain/slasher/service.go [hunk 11]
```diff
@@ -134,7 +135,6 @@ func (s *Service) run() {
 
 // Stop the slasher service.
 func (s *Service) Stop() error {
-	s.cancel()
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:137]: Can we revert this change? I'm worried about long running processes not terminating on shutdown.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

**Review by prestonvanloon — DISMISSED** [review]

**Review by prestonvanloon — APPROVED** [review]
